src/sp/it/pl/main/Tmp.kt

```diff
@@ -1,14 +1,224 @@
+import sp.it.pl.util.conf.IsConfig
+import sp.it.pl.util.conf.MainConfiguration
+import sp.it.pl.util.conf.ValueConfig
+import sp.it.pl.util.conf.computeConfigGroup
+import sp.it.pl.util.conf.obtainConfigConstraints
+import sp.it.pl.util.dev.throwIf
+import sp.it.pl.util.functional.ifNull
+import sp.it.pl.util.type.Util.getGenericPropertyType
+import sp.it.pl.util.validation.Constraint
+import sp.it.pl.util.validation.Constraint.NumberMinMax
+import java.io.File
+import java.util.function.Consumer
+import kotlin.properties.ReadOnlyProperty
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KMutableProperty
+import kotlin.reflect.KProperty
+import kotlin.reflect.full.findAnnotation
+import kotlin.reflect.jvm.isAccessible
+import kotlin.reflect.jvm.javaField
+import kotlin.reflect.jvm.javaType
 
 fun futureWrap(data: Any?): Fut<*> = data as? Fut<*> ?: Fut.fut(data)!!
 
 inline fun <reified T> ActionPane.register(vararg actions: ActionPane.ActionData<T, *>) = register(T::class.java, *actions)
 
 infix fun FXMLController.onDispose(s: () -> Unit) = d({ s() })
 
-infix fun FXMLController.initClose(s: () -> Subscription) = d(s())
\ No newline at end of file
+infix fun FXMLController.initClose(s: () -> Subscription) = d(s())
+
+fun <T: Any> ClassController.v(initialValue: T, onChange: (T) -> Unit) = VLate(initialValue, onChange)
+
+class VLate<T>(private val initialValue: T, private val onChange: (T) -> Unit) {
+    operator fun provideDelegate(ref: ClassController, property: KProperty<*>): ReadOnlyProperty<ClassController, V<T>> {
+        val v = v(initialValue) { if (ref.isInitialized) onChange(it) }
+        return object: ReadOnlyProperty<ClassController, V<T>> {
+            override fun getValue(thisRef: ClassController, property: KProperty<*>): V<T> {
+                return v
+            }
+        }
+    }
+}
+
+fun <T> Config<T>.isEditableByUser() = isEditable.isByUser && constraints.asSequence().none { it is Constraint.ReadOnlyIf && it.condition()}
+
+fun <T: Any> c(initialValue: T): ConfS<T> = ConfS(initialValue).but(Constraint.ObjectNonNull())
+fun <T: Any?> cn(initialValue: T?): ConfS<T?> = ConfS(initialValue)
+fun <T: Any> cv(initialValue: T): ConfV<T, V<T>> = ConfV<T,V<T>>(initialValue).but(Constraint.ObjectNonNull())
+fun <T: Any, W: WritableValue<T>> cv(initialValue: T, valueSupplier: (T) -> W): ConfV<T, W> = ConfV(initialValue, valueSupplier).but(Constraint.ObjectNonNull())
+fun <T: Any?> cvn(initialValue: T?): ConfV<T?, V<T?>> = ConfV(initialValue)
+fun <T: Any?, W: WritableValue<T?>> cvn(initialValue: T?, valueSupplier: (T?) -> W): ConfV<T?, W> = ConfV(initialValue, valueSupplier)
+fun <T: () -> Unit> cr(action: T): ConfR<T> = ConfR(action)
+inline fun <reified T: Any> cList(): ConfL<T> = ConfL(T::class.java)
+
+fun <T: Any?, C: Conf<T>> C.but(vararg restrictions: Constraint<T>) = apply { constraints += restrictions }
+fun <T: Number, C: Conf<T>> C.min(min: T) = but(NumberMinMax(min.toDouble(), Double.MAX_VALUE))
```

Comment:
Are you pointing out potential conflict of constranits MinMax and Max/Min ? Should this be checked and exception thrown?

Reply:
When creating a min, you could check whether a max already exists and replace Double.MAX with it. It might be useful at some point to override previously set limits.